crawl-ref/source/stairs.cc

```diff
@@ -529,6 +552,26 @@ static level_id _travel_destination(const dungeon_feature_type how,
 }
 
 /**
+ * Check to see if the level transition should be stopped
+ *
+ * @param dest      The destination level (branch and depth).
+ * @param feat      The dungeon feature the player is standing on.
+ * @param going_up  True if the player is trying to go up stairs.
+ * @return          True if the level transition should be stopped.
+ */
+static bool _stop_transition(level_id dest, dungeon_feature_type feat,
+                             bool going_up)
+{
+    bool trying_to_win = feat == DNGN_EXIT_DUNGEON && going_up;
+
+    // When actually leaving the dungeon, dest is invalid, so if the player is
+    // trying to win, but the destination is valid, something caused the player
+    // to remain in the dungeon (i.e. slipped back downstairs while confused).
+    return (!dest.is_valid() && !trying_to_win) ||
+            (dest.is_valid() &&  trying_to_win);
```

Comment:
This setup, with the function + comment is _much_ better than the version I accidentally commented on a minute ago :).  I think the XNOR/iff logic now makes sense to me.

Maybe the bool should be named trying_to_exit, since you can escape without the orb.  Alternatively to the return condition here, you could also just return  `dest.is_valid() == trying_to_exit`.  One final thing to consider is if it would be clearer to factor the implicit negation in `stop` (to be explicit negation in take_stairs), and change the helper function to be something like `_level_transition_moves_player`, with the inverted semantics of what it has now.

Reply:
great suggestions. thanks

Reply:
the logic is much simpler now.